src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+                        continue;
+                    }
+                    let qname = question.qname();
+                    let qclass = question.qclass();
+                    let qtype = question.qtype();
+
+                    if !(opcode == Opcode::Query && qclass == Class::In) {
+                        // Anything other than a query on the Internet class
+                        // should not be cached.
+                        let request = self
+                            .upstream
+                            .send_request(self.request_msg.clone());
+                        self.state =
+                            RequestState::GetResponseNoCache(request);
+                        continue;
+                    }
+
+                    let mut ad = header.ad();
+                    let cd = header.cd();
+                    let rd = header.rd();
+
+                    let dnssec_ok = if let Some(opt) = msg.opt() {
+                        opt.dnssec_ok()
+                    } else {
+                        false
+                    };
+                    if dnssec_ok && !ad {
+                        ad = true;
+                    }
+
+                    let key =
+                        Key::new(qname, qclass, qtype, ad, cd, dnssec_ok, rd);
+                    let opt_ce = self.cache_lookup(&key).await?;
+                    if let Some(value) = opt_ce {
+                        let opt_response =
+                            handle_cache_value::<_, C>(qname, value);
+                        if let Some(response) = opt_response {
+                            return response;
+                        }
+                    }
+
+                    let request =
+                        self.upstream.send_request(self.request_msg.clone());
+                    self.state = RequestState::GetResponse(key, request);
+                    continue;
+                }
+                RequestState::GetResponse(key, request) => {
+                    let response = request.get_response().await;
+
+                    let key = key.clone();
+                    let value = Arc::new(Value::new(
+                        &response,
+                        &self.config,
+                        &self.clock,
+                    )?);
+                    self.cache_insert(key, value).await;
+
+                    return response;
+                }
+                RequestState::GetResponseNoCache(request) => {
+                    return request.get_response().await;
+                }
+            }
+        }
+    }
+
+    /// Try to find an cache entry for the key.
+    async fn cache_lookup(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // There are 4 flags that may affect the response to a query.
+        // In some cases the response to one value of a flag could be
+        // used for the ohter value.
+        // This function takes all 4 flags. First we fix the CD flag.
```

Comment:
I'm confused. This function doesn't "fix" anything and "fixing" the CD flag and using it "as is" sound mutually exclusive?

Reply:
I rewrote the comment as:        
        // This function takes all 4 flags. First we take care of the CD flag.
        // This flag has to be used as is, so there is not much to do. Next
        // we pass the request to a function that looks at RD, DO, and AD.